### ISSUE TITLE
Attribute groups

### ISSUE THREAD

**Issue by jetuk** [hydraplatform/hydra-base#10]
We discussed the idea of having some sort of grouping ability on `Attr` objects in the database. This would be useful for organising different types of attribute in both the frontend and backend. There might be two uses cases I guess. Frontend grouping for display, and backend grouping for processing (effectively hidden from the user). Would they be the same field?

**Comment by knoxsp** [comment]
Would it not be more sensible to group ResourceAttr? That way, an attr
remains as a general definition, but a resource attr can be assigned to a
particular grouping within
the context of a specific resource or at the network level?

So in pseudocode, it would look like:

attrs = [attr_1, attr_2, attr_3]


Network
    AttributeGroups
        Group_1 [attr_1]
        Group_2 [attr_2, attr_3]
    Nodes
        Node 1
            attr_1, is_var='Y'
            attr_2, is-var='Y'

Notice the grouping is done at network level

OR

Network
    AttributeGroups
        Group_1 [attr_1]
        Group_2 [attr_2, attr_3]
    Nodes
        Node 1
            attr_1, is_var='Y', groups = [group A]
            attr_2, is-var='Y', groups = [group B]

Here the grouping is done at resource level

**Comment by jetuk** [comment]
You are probably correct. I think I was thinking something like the first example. I have two broad categories of types of "attribute". I want to be able to easily filter to get all network level data/attributes of one category. I think in my example I only ever have a 1:1 mapping between `Attr` and `ResourceAttr`.

What are `group A` and `group B` in the second example? Are they different to `Group_1` and `Group_2`?

**Comment by knoxsp** [comment]
Apologies, that was a copy-paste error
the second one was supposed to read:
Network
    AttributeGroups
        Group_1
        Group_2
    Nodes
        Node 1
            attr_1, is_var='Y', groups = [group 1]
            attr_2, is-var='Y', groups = [group 2]
            attr_3, is-var='Y', groups = [group 2]

Notice that the group contents are not defined in the group definitions,
but in the resource attribute definition.

Steve

**Comment by jetuk** [comment]
OK. This is the structure for the JSON request? Presumably underneath there's just a many-to-many relationship between AttributeGroup and ResourceAttr? You would then be able to find all ResourceAttr for a group or vice-versa?

**Comment by knoxsp** [comment]
Yes, this is an pseudo-code example structure of the json structure. Which
design do you favour? A network-level registry or a resource level?

The db would have a many-to-many relationship, yes.

**Comment by jetuk** [comment]
Probably the second. It seems more logical to define the group on an attribute?

**Comment by jetuk** [comment]
It would be nice to get this done before any more work on the pywr app. Otherwise that app must rely on `Dataset` metadata. However, I feel like we should convert the existing attribute tests to pytest before adding more complexity to the attributes.

**Comment by knoxsp** [comment]
Separate issue created. As it's a blocker for the PywrApp, i'll get on it
ASAP.
S

**Comment by knoxsp** [comment]
We discussed this further today. I think the summary is as follows,

- The groups will be `Attr` groups with a table `tAttrGroup`.
- The mapping between these groups and specific attributes will be done at network level.
- Therefore this creates an association table between a network, attribute group and attribute.
- The above table will have a constraint to ensure that an attribute can only in a single group in a given network.

None of this touches resource attributes, etc.

One concern was the global nature of the definition of attributes themselves, and whether any of the above can create side effects between projects and networks. For example removing an attribute which is used by two different networks (i.e. they coincidentally had the same name). Having the attribute grouping at network level offsets this somewhat.
